### PR TITLE
Add TypeScript declarations

### PR DESCRIPTION
**Pull request by kasperisager**
This pull request intends to add TypeScript declarations for Brittle. Initially, the declarations have been autogenerated based on the inference capabilities of TypeScript. The only thing currently inferred by TypeScript are the names of the top-level exports so this approach won't produce useful declarations unless the JavaScript source is also adjusted. The other option is to author the type declarations by hand, which of course means that any change to public API will have to be manually reflected in the type declarations.

To summarise, I see two approaches to introducing TypeScript declarations:

1. Adjust the JavaScript source to improve type inference, adding JSDoc annotations as needed (https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html).

2. Write the declarations by hand.

Which one would be preferred?

### CODE REVIEW

**Comment by davidmarkclements** [comment]
I think let's go by hand for now. While it's true that api modifications will have to be updated in dts files by hand, it's also true that jsdoc headers would have to be updated by hand as well. 

We can catch API differences that may occur with an approach similar to that used in pino (see https://github.com/pinojs/pino/blob/master/package.json#L30 and https://github.com/pinojs/pino/blob/master/test/types/pino-import.test-d.ts)

**Comment by kasperisager** [comment]
I think I've now covered the entire public API so this should be ready for another look.

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+  error: Error | null
+  ended: boolean
+}
+
+declare interface Test extends Promise<Metadata>, Assertions, Utilities, Metadata {
+  test: TestFn
+  skip: TestFn
+  todo: TestFn
+  configure(options: TestOptions): void
+}
```
Comment (davidmarkclements) [inline index.d.ts:43]: the `test` instance also has a circular reference to itself, would this work?

```suggestion
  assert: Test
}
```

File: index.d.ts [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+declare interface TestOptions {
+  /**
+   * @default 30000
+   */
+  timeout?: number
+
+  /**
+   * @default process.stderr
```
Comment (davidmarkclements) [inline index.d.ts:57]: `output` can be a stream or a number (representing a file descriptor). It defaults to `1` (the fd for stdout)

File: index.d.ts [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+}
+
+declare interface Metadata {
+  start: bigint
+  description: string
+  planned: number
+  count: number
+  error: Error | null
+  ended: boolean
+}
```
Comment (davidmarkclements) [inline index.d.ts:36]: there's also:

index, passing, failing, time - all numbers (time is decimal)

File: test/index.test-d.ts [hunk 4]
```diff
@@ -0,0 +1,8 @@
+import { expectType } from 'tsd'
+import test, { configure } from '../index.js'
+
+configure({ serial: true })
+
+test('some test', async (assert) => {
+  expectType<boolean>(await assert.is(true, true))
+})
```
Comment (davidmarkclements) [inline test/index.test-d.ts:8]: can we check the types of exports (similar to https://github.com/pinojs/pino/blob/master/test/types/pino-import.test-d.ts but within a `test` as it is already, is great), and the types on `assert` - the resolved boolean values I'm less bothered about but we can check those as well

**Comment by kasperisager** [comment]
I've adjusted the types and tests and also added the coercible assertions, which I realised were missing.

File: index.d.ts [hunk 5]
```diff
@@ -0,0 +1,109 @@
+import { WriteStream } from "fs"
+
+interface Coercable {
```
Comment (davidmarkclements) [inline index.d.ts:3]: I think the name of this overly emphasises coercion, can we maybe call it something like `Comparator` ?
e.g. a Comparator is an assertion that compares two values, strictly or coercively

File: test/index.test-d.ts [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+  Metadata,
+  test,
+  skip,
+  todo,
+  configure
+} from '../index.js'
+
+// Top-level exports
+
+expectType<TestFn>(brittle)
```
[the remark this answers is not in the document]
Comment (davidmarkclements) [reply test/index.test-d.ts:14]: ah right right, understood

**Review by davidmarkclements — APPROVED** [review]
excellent - thank you!
